src/Interpreters/Cache/Metadata.cpp

```diff
@@ -536,13 +536,23 @@ void CacheMetadata::downloadThreadFunc()
                 holder = std::make_unique<FileSegmentsHolder>(FileSegments{file_segment});
             }
 
-            downloadImpl(holder->front(), memory);
+            auto & file_segment = holder->front();
+
+            if (file_segment.getOrSetDownloader() != FileSegment::getCallerId())
+                continue;
+
+            chassert(file_segment.getDownloadedSize() != file_segment.range().size());
+            chassert(file_segment.assertCorrectness());
+
+            downloadImpl(file_segment, memory);
         }
         catch (...)
         {
             if (holder)
             {
-                const auto & file_segment = holder->front();
+                auto & file_segment = holder->front();
+                file_segment.setDownloadFailed();
```

Comment:
The main fix is in this line. It prevents submitting the file segment back to download queue in `~FileSegmentsHolder() -> FileSegment::complete()`